### PR TITLE
Add switch and climate entities, integrate with DataUpdateCoordinator

### PR DESCRIPTION
**Pull request by lexakalex**
Please let me know what do you think about the general approach and I can split the PR into multiple smaller ones to make it easier to review.

1. I slightly refactored the original water heater and binary sensors and added more devices.
3. Integrated all entities with DataUpdateCoordinator. This allows for a central place to store Kohler state and a provides a listener mechanism to notify entities about changes. Without this, actions were somewhat laggy. For example, when I turn an outlet off via a switch, it would take seconds for the change to propagate into climate / water heater / binary sensor. Now it's almost immediate.
4. Added switch entities to allow turning individual outlets on / off
5. Added a climate entity so that I can use a thermostat card

### CODE REVIEW

**Review by niemyjski — COMMENTED** [review]
Thanks for the PR! I think it would be great to update pykohler with any changes and then we can just use the library for request which is needed if we ever merge this into HA Core.

File: custom_components/kohler/kohler.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
+import json
+import requests
+from requests.exceptions import ConnectionError
+import logging
+
+from .const import CONF_ACCEPT_LIABILITY_TERMS, DOMAIN, DATA_KOHLER
+
+_LOGGER = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (lexakalex) [reply custom_components/kohler/kohler.py:8]: sounds good. will extract into a separate PR into that repo.

**Comment by lexakalex** [comment]
What are your thoughts on moving over to DataUpdateCoordinator?
I use it to propagate changes to all entities at once rather than waiting for entities themselves to poll for changes.

I wonder if you yourself experienced any lag when turning shower on via water heater entity and then waiting for binary sensors to catch up.

If I were to work on mering this logic in, I'm thinking the way to go would be 4 PRs:
0. Move kohler.py changes to your other repo. (already agreed)
1. Refactor the current code to use DataUpdateCoordinator, keep everything else the same
2. Add support for switch entity (requires minor changes to binary sensor)
3. Add support for thermostat entity (while keeping water heater for backwards compat)

**Comment by lexakalex** [comment]
https://github.com/user-attachments/assets/24a1c3be-6fce-46fe-8df7-87bb726ab5fe

**Comment by niemyjski** [comment]
Changes look awesome. There has always been a delay due to polling. And I do notice the lag from time to time! If we can be more efficient like this it's a massive win and I greatly appreciate the help and this pr

**Comment by niemyjski** [comment]
0.0.6 has been pushed (https://github.com/niemyjski/kohler-python/actions/runs/11601718659/job/32305110525) after a few attempts at creating the workflow 😅

**Comment by dcmeglio** [comment]
Seeing an issue I haven't tracked down yet. It's reporting outlet 6 is on but it isn't. When I revert your pr it correctly shows output 6 is off so this seems to be introduced in this pr

**Comment by niemyjski** [comment]
@lexakalex any luck on tracking down the valve issue? Thanks again for all of your help

**Comment by niemyjski** [comment]
@dcmeglio looking at the code you posted the api is returning `"valve2outlet6": true,` so it is saying it's on. How many valves do you have installed?

**Comment by niemyjski** [comment]
Thanks for the PR. I'm going to fix up the requests in the branch and then merge that in

**Comment by niemyjski** [comment]
@lexakalex Would it be possible for you to create a follow up pr for the following:

```
Detected that custom integration 'kohler' uses `async_config_entry_first_refresh`, which is only supported for coordinators with a config entry and will stop working in Home Assistant 2025.11 at custom_components/kohler/__init__.py, line 63: await hass.data[DATA_KOHLER].async_config_entry_first_refresh(), please create a bug report at https://github.com/niemyjski/homeassistant-kohler/issues
```

**Comment by niemyjski** [comment]
image
 I'm also seeing some outlets on/off , we might need to investigate why the controller does this.

**Comment by dcmeglio** [comment]
> image I'm also seeing some outlets on/off , we might need to investigate why the controller does this.

Check with https://github.com/niemyjski/homeassistant-kohler/pull/34 this is working for me now.

**Comment by dcmeglio** [comment]
> Timeout fetching Kohler Data Coordinator data

I experienced this issue too. I haven't investigated what's going on just yet though

> Detected that custom integration 'kohler' uses `async_config_entry_first_refresh`, which is only supported for coordinators with a config entry and will stop working in Home Assistant 2025.11 at custom_components/kohler/__init__.py, line 63: await hass.data[DATA_KOHLER].async_config_entry_first_refresh(), please create a bug report at https://github.com/niemyjski/homeassistant-kohler/issues
This one's not occurring for me

**Pull request merged**